### ISSUE TITLE
haddocks incorrectly parses UNPACK pragma

### ISSUE THREAD

**Issue by chessai** [haskell/haddock#836]
This might be a problem with cabal new-haddock, but given that it fails in a parsing stage, I think it's haddock causing the problem. Let me know if you think I should open up the issue on cabal as well/instead.

```bash
#GHC version
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.4.2

#cabal version
$ cabal --version
cabal-install version 2.2.0.0
compiled using version 2.2.0.1 of the Cabal library

#haddock version
$ haddock --version
Haddock version 2.20.0, (c) Simon Marlow 2006
Ported to use the GHC API by David Waern 2006-2008

#Operating System
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.4 LTS
Release:	16.04
Codename:	xenial

#Replicate
$ git clone git@github.com:chessai/freq && cd freq
freq$ nix-shell
nix-shell$ cabal new-build #this will succeed
nix-shell$ rm -rf dist*
nix-shell$ cabal new-haddock #this will fail with the following:
Build profile: -w ghc-8.4.2 -O1
In order, the following will be built (use -v for more details):
 - freq-0.1.0.0 (lib) (first run)
Preprocessing library for freq-0.1.0.0..
Running Haddock on library for freq-0.1.0.0..
Haddock coverage:

src/Freq/Internal.hs:208:3: error:
    parse error on input ‘{-# UNPACK’
    |
208 |   {-# UNPACK #-} !ByteArray -- ^ Square two-dimensional array of Double, maps first char and second char to probability
    |   ^^^^^^^^^^
cabal: Failed to build documentation for freq-0.1.0.0.

```

Most of the above was inside of a nix-shell.
I tried to create a minimal example of just a module with a similarly laid-out (textually) data type, like so:

```haskell
{-# LANGUAGE BangPatterns #-}

module Example
  ( -- * The Foo type
    Foo(..)
  ) where

data Foo = Foo
  {-# UNPACK #-} !Int
  {-# UNPACK #-} ![Int]

```

but haddock parses this correctly, under the same setup, inside of the same nix shell.

**Comment by chessai** [comment]
OK, thank you

**Comment by harpocrates** [comment]
Note that the GHC side of this should be now fixed in 8.6 too (https://github.com/ghc/ghc/commit/0361fc038e117befc3c59fcd589d640006407ed6)!

**Comment by chessai** [comment]
Thank you @harpocrates !
